xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp

```diff
@@ -975,11 +976,24 @@ bool CDVDVideoCodecAndroidMediaCodec::AddData(const DemuxPacket &packet)
                                          packet.cryptoInfo->numSubSamples);
 
         cryptoInfo = new CJNIMediaCodecCryptoInfo();
+        if (CJNIBase::GetSDKVersion() < 25 &&
+            packet.cryptoInfo->mode == CJNIMediaCodec::CRYPTO_MODE_AES_CBC)
+        {
+          CLog::LogF(LOGERROR, "Device API does not support CBCS decryption");
+          return false;
+        }
 
         cryptoInfo->set(packet.cryptoInfo->numSubSamples, clearBytes, cipherBytes,
                         std::vector<char>(packet.cryptoInfo->kid, packet.cryptoInfo->kid + 16),
                         std::vector<char>(packet.cryptoInfo->iv, packet.cryptoInfo->iv + 16),
-                        CJNIMediaCodec::CRYPTO_MODE_AES_CTR);
+                        packet.cryptoInfo->mode == CJNIMediaCodec::CRYPTO_MODE_AES_CBC
+                            ? CJNIMediaCodec::CRYPTO_MODE_AES_CBC
+                            : CJNIMediaCodec::CRYPTO_MODE_AES_CTR);
+
+        cryptoInfoPattern =
+            std::make_unique<CJNIMediaCodecCryptoInfoPattern>(CJNIMediaCodecCryptoInfoPattern(
+                packet.cryptoInfo->cryptBlocks, packet.cryptoInfo->skipBlocks));
+        cryptoInfo->setPattern(*cryptoInfoPattern.get());
```

Comment:
This doesn't do what you think it does. When you create a unique_ptr, it owns the object, and when the unique_ptr falls out of scope (as it does here), the object is deleted.

A good hint is to scrutinize when an underlying pointer is exposed. `.get()` should be avoided. Indeed, if you did `*cryptoInfoPattern` it would have the exact same effect as `*cryptoInfoPattern.get()` and avoid the `.get()`. If we require the underlying pointer, then it's best to call `release()` and delegate lifetime to whatever pointer management follows. If `release()` doesn't work, then ownership can't be released, and the pointer will have to persist past the local scope.

I would expect to see here `cryptoInfo->setPattern(std::move(cryptoInfoPattern));`, not sure if that's possible though.

Reply:
I just noticed that `cryptoInfo` is in the parent scope by a single block, so maybe this is OK. Nevertheless it deserves scrutinization.

Reply:
`cryptoInfo->setPattern(std::move(cryptoInfoPattern));` isn't giving me success - libandroidjni really just wants `const CJNIMediaCodecCryptoInfoPattern`
Is this ok to proceed as-is or do I need to continue searching for another way around?

Reply:
I haven't checked the code of cryptoInfo, but you should be fine with `cryptoInfo->setPattern(*cryptoInfoPattern);` as long as the lifetime of `cryptoInfoPattern` is sufficient. Possibly by promoting to a member variable? If you can't find a good lifetime for the variable I'll look into it and figure out what to do.

Reply:
I'm guessing that `cryptoInfoPattern` is copy-constructed, so no need to use the heap via unique_ptr. Does the following work?

```c++
CJNIMediaCodecCryptoInfoPattern cryptoInfoPattern(packet.cryptoInfo->cryptBlocks, packet.cryptoInfo->skipBlocks));
cryptoInfo->setPattern(cryptoInfoPattern);
```

Reply:
Yes it does, thank you!